### PR TITLE
Inductor allgather_into_tensor

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95530

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by kumpera — APPROVED** [review]
LGTM.

File: torch/distributed/_functional_collectives.py [hunk 1]
```diff
@@ -160,6 +160,20 @@ def _all_reduce(self, reduceOp, tag, ranks, group_size):
 c10_lib_cpu.impl("wait_tensor", _wait_tensor)
 c10_lib_cuda.impl("wait_tensor", _wait_tensor)
 
+def _all_gather_into_tensor(shard, tag, ranks, group_size):
+    # TODO add dim support?
+    group = c10d._find_or_create_pg_by_ranks_and_tag(tag, ranks, group_size)
+    assert group is not None
+    out_size = list(shard.size())
+    out_size[0] *= group_size
+    out_tensor = shard.new_empty(out_size)
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/_functional_collectives.py:169]: yes, i checked.

**Pull request closed without merging**